src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -2145,9 +2145,13 @@ impl Coordinator {
             mut optimizer,
         }: PeekStageOptimize,
     ) -> Result<PeekStageRealTimeRecency, AdapterError> {
-        let local_mir_plan = optimizer.optimize(source)?;
-        let local_mir_plan = local_mir_plan.resolve(session, stats);
-        let global_mir_plan = optimizer.optimize(local_mir_plan)?;
+        let global_mir_plan = catch_unwind(true, "peek", || {
+            let local_mir_plan = optimizer.optimize(source)?;
```

Comment:
Overall I like this change, the fewer chances `environmentd` panics the better!

From a design perspective I'm wondering if it might be nice to put the `catch_unwind` inside of the `Optimize` trait? e.g. something like:

```
// private trait
trait OptimizeInternal {
  fn optimize_plan(...) -> Result<_, OptimizerError>
}

// public trait
pub trait Optimize {
  fn optimize(...) -> Result<_, OptimizeError>;
}

impl<T: OptimizeInternal> Optimize for T {
  fn optimize(...) -> Result<_, OptimizerError> {
    catch_unwind(|| self.optimize_plan(...))
  }
}
```
this way we never forget to wrap a call to optimize in `catch_unwind`?

Reply:
I like that. I'll extend the `Optimize` API tomorrow as proposed.

Reply:
Sounds good to me!

Reply:
Finally I just added a second method `catch_unwind_optimize` with default implementation as you proposed in the `Optimize` trait. Once I fix off-thread optimization for the `sequence_explain_~` paths I might reconsider the split between public and internal traits, but for the moment I still need the `optimize` calls in the `explain_~` methods.